File: Data Structures & Algorithms/max-area-of-island/submission-0.java
class Solution {
      int[][] dir = {{1,0},{-1,0},{0,1},{0,-1}};
    public int maxAreaOfIsland(int[][] grid) {
        int max =0 ;
      
        int n  = grid.length,m=grid[0].length;
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]==1){
                    max = Math.max(dp(grid,i,j),max);
                }
            }
        }
        return max;
    }
    private int dp(int[][] grid , int row,int col){
            if(row>=grid.length || row<0 || col>=grid[0].length || col<0 || grid[row][col]==0){
                return 0;
            }
            int res=1;
            grid[row][col]=0;
            for(int[] d: dir){
                res+=dp(grid,row+d[0],col+d[1]);
            }
            return res;
        }
    
}

File: Data Structures & Algorithms/swim-in-rising-water/submission-0.cpp
class Solution {
public:
   int n;
   vector<vector<int>> dir = {{1,0},{-1,0},{0,-1},{0,1}};
    bool check(vector<vector<int>>& grid, int i,int j,int t, vector<vector<bool>>& visited){
        if(i>=n || i<0 || j>=n || j<0 || visited[i][j]==true || grid[i][j]>t)return false;
        visited[i][j]=true;
        if(i==n-1 && j==n-1)return true;

        for(auto &it : dir){
            int i_ = i+it[0];
            int j_ = j+it[1];
            if(check(grid,i_,j_,t,visited))return true;
        }
        return false;
    }
    int swimInWater(vector<vector<int>>& grid) {
         n = grid.size();
        int l = grid[0][0];
        int h=n*n;
        // for(int i=0;i<n;i++){
        //     for(int j=0;j<grid[0].size();j++){
        //         h = max(grid[i][j],h);
        //     }
        // }
        int res=0;
        while(l<=h){
            int mid = (l+h)/2;
            vector<vector<bool>> visited(n,vector<bool>(n,false));
            if(check(grid,0,0,mid,visited)){
                res=mid;
                h=mid-1;
            }
            else{
                l=mid+1;
            }
        }
        return res;
    }
};
